### PR TITLE
Template support

### PR DESCRIPTION
**Pull request by sanderPostma**

### CODE REVIEW

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts [hunk 1]
```diff
@@ -170,17 +173,16 @@ const credentialDataSupplierEnergySHRGuest2023: CredentialDataSupplier = (args:
...
     const lastName = args.credentialDataSupplierInput?.lastName ?? 'Sphereon'
     const email = args.credentialDataSupplierInput?.email ?? 'sphereon@example.com'
 
-    if (args.credentialRequest.format !== 'jwt_vc_json') {
-        throw Error(`Format ${args.credentialRequest.format} is not configured on this issuer`)
-    }
-
     const types = getTypesFromRequest(args.credentialRequest)
     if (types.includes('VerifiedEmployee')) {
+        if (args.credentialRequest.format !== 'jwt_vc_json') {
```
Comment (nklomp) [inline packages/agent/src/utils/oid4vciCredentialSuppliers.ts:183]: Why is this added? Whether it is supported really depends on the conf files

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts [hunk 2]
```diff
@@ -198,6 +200,9 @@ const credentialDataSupplierSphereon: CredentialDataSupplier = (args: Credential
             },
         } as unknown as CredentialDataSupplierResult)
     } else if (types.includes('MembershipExample')) {
+        if (args.credentialRequest.format !== 'jwt_vc_json') {
```
Comment (nklomp) [inline packages/agent/src/utils/oid4vciCredentialSuppliers.ts:203]: Why is this added? Whether it is supported really depends on the conf files

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts [hunk 3]
```diff
@@ -242,18 +252,66 @@ const supplierByType = async (args: CredentialDataSupplierArgs): Promise<Credent
...
+    }
+
+    public async generateCredentialData(args: CredentialDataSupplierArgs): Promise<CredentialDataSupplierResult> {
+        const hardCodedSupplierResult = await this.getHardCodedSupplierCredential(this.issuerCorrelationId, args)
+        if (hardCodedSupplierResult.credential) {
+            return Promise.resolve(hardCodedSupplierResult)
+        }
+
+        const types: string[] = getTypesFromRequest(args.credentialRequest)
+        const credentialSupplierConfig = args.credentialSupplierConfig as CredentialSupplierConfigWithTemplateSupport
```
Comment (nklomp) [inline packages/agent/src/utils/oid4vciCredentialSuppliers.ts:277]: Why is the as needed?

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts [hunk 4]
```diff
@@ -242,18 +252,66 @@ const supplierByType = async (args: CredentialDataSupplierArgs): Promise<Credent
...
+                }
+                return Promise.resolve({
+                    format: templateMapping.format || args.credentialRequest.format,
+                    credential: credential
+                } as unknown as CredentialDataSupplierResult)
+            } else {
+                throw new Error(`No template mapping could be found for types ${types.join(', ')}`)
+            }
+        }
+        throw new Error(`No supplier credential be found for types ${types.join(', ')} and correlationId ${this.issuerCorrelationId}`)
```
Comment (nklomp) [inline packages/agent/src/utils/oid4vciCredentialSuppliers.ts]: Can this error message be improved a bit

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts [hunk 5]
```diff
@@ -242,18 +252,66 @@ const supplierByType = async (args: CredentialDataSupplierArgs): Promise<Credent
...
+        }
+
+        const types: string[] = getTypesFromRequest(args.credentialRequest)
+        const credentialSupplierConfig = args.credentialSupplierConfig as CredentialSupplierConfigWithTemplateSupport
+        if (credentialSupplierConfig.template_mappings) {
+            const templateMapping = credentialSupplierConfig.template_mappings
+                .find(mapping => mapping.credential_types.some(type => types.includes(type)))
+            if (templateMapping) {
+                const baseDir = credentialSupplierConfig?.templates_base_dir ? `/${credentialSupplierConfig.templates_base_dir}` : '';
+                const templatePath = `${CONF_PATH}${baseDir}/${templateMapping.template_path}`;
```
Comment (nklomp) [inline packages/agent/src/utils/oid4vciCredentialSuppliers.ts]: I rather see a bit of logic looking at slashes when concatenating. You don't want to end up with 'folder1//folder2' or 'folder1folder2`. There are 3 vars here so it could go wrong for all three of them

**Review by nklomp — CHANGES_REQUESTED** [review]
See some small remarks.

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 1 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:183]: That code was there, it was just moved from outside the  if (types.includes('VerifiedEmployee')) { scope to within

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 2 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:203]: That code was there, it was just moved from outside the  if (types.includes('MembershipExample')) { scope to within

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 3 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:277]: Code completion, CredentialSupplierConfig has type [key: string]: any; in the child project
image

Comment (nklomp) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:277]: That screenshot doesn't show why it is needed. In other words isn't it already typed?

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 5 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts]: Added function normalizeFilePath

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 4 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts]: done

**Comment by sanderPostma** [comment]
hbs templates have been updated to get all opers from the inputData and dump them in credentialSubject

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts [hunk 6]
```diff
@@ -226,9 +236,9 @@ const credentialDataSupplierSphereon: CredentialDataSupplier = (args: Credential
                     type: 'Sphereon Guest',
                 },
             },
-        } as unknown as CredentialDataSupplierResult)
-    }
-    throw Error(`${JSON.stringify(types)} not supported by this issuer`)
+        } as unknown as CredentialDataSupplierResult
+    } */
+    return Promise.resolve({} as unknown as CredentialDataSupplierResult) // return empty result so the code will start looking for templates
```
Comment (nklomp) [inline packages/agent/src/utils/oid4vciCredentialSuppliers.ts:241]: Please resolve to undefined instead of an empty object

File: packages/agent/conf/examples/templates/kvk/KVKRegistration.hbs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+    "@context": [
+        "https://www.w3.org/2018/credentials/v1"
+    ],
+    "type": [
+        "VerifiableCredential",
+        "KVKRegistration"
+    ],
+    "description": "This credential proves that the company you represent is registered with KVK",
```
Comment (nklomp) [inline packages/agent/conf/examples/templates/kvk/KVKRegistration.hbs]: Description should be in the credential subject and made shorter. We need to be able to display it.

"Company registered with the KvK"

File: packages/agent/conf/examples/templates/auckland/NotaryCompanyVerification.hbs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
+    "@context": [
+        "https://www.w3.org/2018/credentials/v1"
+    ],
+    "type": [
+        "VerifiableCredential",
+        "NotaryCompanyVerification"
+    ],
+    "description": "This credential proves that the company you represent is registered with the Van Auckland Notary",
```
Comment (nklomp) [inline packages/agent/conf/examples/templates/auckland/NotaryCompanyVerification.hbs]: Description should be in the credential subject and made shorter. We need to be able to display it.

"Company has been verified by the notary"

**Review by nklomp — CHANGES_REQUESTED** [review]
Some small remarks. Also the PR has conflicts currently

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 3 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:277]: CredentialSupplierConfig has type [key: string]: any; It's in OID4VCI and loaded with an isolated section in the issuer metadata (credential_supplier_config)
OID4VCI does not know what that config is going to be, that's up to the implementing project.

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 6 above)
Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:241]: I have tried but it's complicated, the function has to implement CredentialDataSupplier which is defined in dependency OID4VC
```
export type CredentialDataSupplier = (args: CredentialDataSupplierArgs) => Promise<CredentialDataSupplierResult>
```
It is not Promise<CredentialDataSupplierResult | undefined>

File: packages/agent/conf/examples/templates/kvk/KVKRegistration.hbs (hunk 7 above)
Comment (sanderPostma) [reply packages/agent/conf/examples/templates/kvk/KVKRegistration.hbs]: done

File: packages/agent/conf/examples/templates/auckland/NotaryCompanyVerification.hbs (hunk 8 above)
Comment (sanderPostma) [reply packages/agent/conf/examples/templates/auckland/NotaryCompanyVerification.hbs]: done

File: packages/agent/src/utils/oid4vciCredentialSuppliers.ts (hunk 6 above)
Comment (nklomp) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:241]: Okay then create a TODO in the code here. We should move all the current hardcoded stuff to templates anyway, so this whole logic can go

Comment (sanderPostma) [reply packages/agent/src/utils/oid4vciCredentialSuppliers.ts:241]: done

**Comment by nklomp** [comment]
LGTM
